File: server/src/hsogserver/utils/geometry/Line2D.cpp
#include "Line2D.h"
#include "Angle.h"
#include "Geometry.h"
#include "Pose2D.h"

#include <boost/concept_check.hpp>
#include <cmath>
#include <limits>

using namespace taco;
using namespace Eigen;

Line2D::Line2D() : _start(0, 0), _extension(1, 0)
{
}

Line2D::Line2D(const double &m, const double &b) : _start(0, b), _extension(1, m)
{
}

Line2D::Line2D(const double &xStart, const double &yStart, const Angle &alpha, const double &length)
	: _start(xStart, yStart), _extension(alpha.getVector(length))
{
	if (_extension(0) == 0 && _extension(1) == 0) {
		_extension = Vector2d(1, 0);
	}
}

Line2D::Line2D(const Eigen::Vector2d &start, const Angle &alpha, const double &length)
	: _start(start), _extension(alpha.getVector(length))
{
	validate();
}

Line2D::Line2D(const Vector2d &start, const Vector2d &end) : _start(start), _extension(end - start)
{
	validate();
}

Line2D::Line2D(const double &xStart, const double &yStart, const double &xEnd, const double &yEnd)
	: _start(xStart, yStart), _extension(xEnd - xStart, yEnd - yStart)
{
	validate();
}

Line2D::~Line2D()
{
}

void Line2D::validate()
{
	if (_extension(0) == 0 && _extension(1) == 0) {
		_extension = Vector2d(0, 1);
	}
}

const Eigen::Vector2d &Line2D::getStart() const
{
	return _start;
}

const Eigen::Vector2d &Line2D::getExtensionVector() const
{
	return _extension;
}

const Eigen::Vector2d Line2D::getEnd() const
{
	return _start + _extension;
}

const Angle Line2D::getAngle() const
{
	return Angle::to(_extension);
}

const double Line2D::m() const
{
	if (_extension(0) == 0) {
		if (_extension(1) > 0) {
			return std::numeric_limits<double>::max();
		} else {
			return std::numeric_limits<double>::lowest();
		}
	}

	return _extension(1) / _extension(0);
}

const double Line2D::b() const
{
	if (_extension(0) == 0) {
		return 0;
	}

	return _start(1) - ((_extension(1) / _extension(0)) * _start(0));
}

const double Line2D::yValue(const double &x) const
{
	if (_extension(0) == 0) {
		// this line is y-parallel
		return 0;
	}

	return (_extension(1) / _extension(0)) * (x - _start(0)) + _start(1);
}

const double Line2D::xValue(const double &y) const
{
	if (_extension(1) == 0) {
		// this line is x-parallel
		return _start(0);
	}

	return (_extension(0) / _extension(1)) * (y - _start(1)) + _start(0);
}

Pose2D Line2D::getClosestPose(const Vector2d &point)
{
	Angle lineAngle(Angle::to(_extension));
	Angle pointAngle(Angle::to(point - _start));

	Angle offsetAngle(pointAngle - lineAngle);
	double hypotenuse = Geometry::getDistance<double, 2>(point, _start);
	double factor = cos(offsetAngle.rad()) * hypotenuse;
	Vector2d closestPoint(_start + _extension / Geometry::getNorm<double, 2>(_extension) * factor);

	return Pose2D(closestPoint, Angle::to(_extension));
}

std::vector<Vector2d> Line2D::getTrail(const Vector2d &start, const Vector2d &end)
{
	Pose2D poseStart = getClosestPose(start);
	Pose2D poseEnd = getClosestPose(end);
	double distance = poseStart.getDistanceTo(poseEnd);
	Vector2d offset = poseStart.getAngle().getVector(0.1 * distance);
	Vector2d point = poseStart.getPosition();
	std::vector<Vector2d> points;
	int i = 0;
	do {
		point += offset;
		points.push_back(point);
	} while (++i < 10);
	return points;
}

bool Line2D::operator==(const Line2D &other) const
{
	return _start == other._start && _extension == other._extension;
}

bool Line2D::operator!=(const Line2D &other) const
{
	return !(*this == other);
}

File: client/base/src/hso/autonomy/util/geometry/Line2D.java
package hso.autonomy.util.geometry;

import java.util.ArrayList;
import java.util.List;

import org.apache.commons.math3.geometry.euclidean.threed.Vector3D;
import org.apache.commons.math3.geometry.euclidean.twod.Vector2D;

public class Line2D
{
	/** The start point of the line segment. */
	private Vector2D start;

	/** The line segment extension. */
	private Vector2D extension;

	/** Construct a line from point (0, 0) to (1, 0) */
	public Line2D()
	{
		this.start = new Vector2D(0, 0);
		this.extension = new Vector2D(1, 0);
	}

	/** Construct a line with the given slope and y-offset
	 * @param m - the slope of the line
	 * @param b - the y-offset of the line
	 */
	public Line2D(final double m, final double b)
	{
		this.start = new Vector2D(0, b);
		this.extension = new Vector2D(1, m);
	}

	/** Construct a line with a start point, and angle and length of the line.
	 * @param xStart - the x-coordinate of the start point (or an arbitrary point on the line)
	 * @param yStart - the y-coordinate of the start point (or an arbitrary point on the line)
	 * @param alpha - the angle from the x-axis to the line
	 * @param length - the line segment length (or any value > 0)
	 */
	public Line2D(final double xStart, final double yStart, final Angle alpha, final double length)
	{
		this.start = new Vector2D(xStart, yStart);
		this.extension = new Vector2D(Math.cos(alpha.radians()) * length, Math.sin(alpha.radians()) * length);
		validate();
	}

	/** Construct a line with a start point, and angle with a length of 1.
	 * @param xStart - the x-coordinate of the start point (or an arbitrary point on the line)
	 * @param yStart - the y-coordinate of the start point (or an arbitrary point on the line)
	 * @param alpha - the angle from the x-axis to the line
	 */
	public Line2D(final double xStart, final double yStart, final Angle alpha)
	{
		this(xStart, yStart, alpha, 1);
	}

	/** Construct a line with a start point, and angle and length of the line.
	 * @param start - the start point of the line segment (or an arbitrary point on the line)
	 * @param alpha - the angle from the x-axis to the line
	 * @param length - the line segment length (or any value > 0)
	 */
	public Line2D(final Vector2D start, final Angle alpha, final double length)
	{
		this.start = start;
		this.extension = new Vector2D(Math.cos(alpha.radians()) * length, Math.sin(alpha.radians()) * length);
		validate();
	}

	/** Construct a line with a start point, and angle with a length of 1.
	 * @param start - the start point of the line segment (or an arbitrary point on the line)
	 * @param alpha - the angle from the x-axis to the line
	 */
	public Line2D(final Vector2D start, final Angle alpha)
	{
		this(start, alpha, 1);
	}

	/** Construct a line with a start and end point (or two arbitrary points on a the line).
	 * @param start - the start point of the line segment
	 * @param end - the end point of the line segment
	 */
	public Line2D(final Vector2D start, final Vector2D end)
	{
		this.start = start;
		this.extension = end.subtract(start);
		validate();
	}

	/** Construct a line with a start and end point (or two arbitrary points on a the line).
	 * @param xStart - the x-coordinate of the start point of the line segment
	 * @param yStart - the y-coordinate of the start point of the line segment
	 * @param xEnd - the x-coordinate of the end point of the line segment
	 * @param yEnd - the y-coordinate of the end point of the line segment
	 */
	public Line2D(final double xStart, final double yStart, final double xEnd, final double yEnd)
	{
		this.start = new Vector2D(xStart, yStart);
		this.extension = new Vector2D(xEnd - xStart, yEnd - yStart);
	}

	/** Retrieve the start point of the line segment. */
	public Vector2D getStart()
	{
		return this.start;
	}

	/** Retrieve the line segment extension vector. */
	public Vector2D getExtensionVector()
	{
		return this.extension;
	}
	/** Retrieve the end point of the line segment. */
	Vector2D getEnd()
	{
		return this.start.add(this.extension);
	}

	/** Retrieve the angle from the x-axis to this line. */
	public Angle getAngle()
	{
		return Angle.to(this.extension);
	}

	/** Retrieve the slope of this line (of line-formula: y = m*x + b). */
	public double m()
	{
		if (this.extension.getX() == 0) {
			if (this.extension.getY() > 0) {
				return Double.MAX_VALUE;
			} else {
				return Double.MIN_VALUE;
			}
		}

		return this.extension.getY() / this.extension.getX();
	}

	/** Retrieve the y-offset of this line (of line-formula: y = m*x + b). */
	public double b()
	{
		if (this.extension.getX() == 0) {
			return 0;
		}

		return this.start.getY() - ((this.extension.getY() / this.extension.getX()) * this.start.getX());
	}

	/** Retrieve the y-value of this line at the given x-value */
	public double yValue(double x)
	{
		if (this.extension.getX() == 0) {
			// this line is y-parallel
			return 0;
		}

		return (this.extension.getY() / this.extension.getX()) * (x - this.start.getX()) + this.start.getY();
	}

	/** Retrieve the x-value of this line at the given y-value */
	public double xValue(double y)
	{
		if (this.extension.getY() == 0) {
			// this line is x-parallel
			return this.start.getX();
		}

		return (this.extension.getX() / this.extension.getY()) * (y - this.start.getY()) + this.start.getX();
	}

	/** Retrieve the closest point and slope angle of this point on the line from given point */
	public Pose2D getClosestPose(final Vector2D point)
	{
		Angle lineAngle = Angle.to(this.extension);
		Angle pointAngle = Angle.to(point.subtract(this.start));

		Angle offsetAngle = pointAngle.subtract(lineAngle);
		double hypotenuse = point.distance(this.start);
		double factor = Math.cos(offsetAngle.radians()) * hypotenuse;
		Vector2D closestPoint = new Vector2D(this.extension.getX() / this.extension.getNorm() * factor,
				this.extension.getY() / this.extension.getNorm() * factor);
		closestPoint = this.start.add(closestPoint);

		return new Pose2D(closestPoint, Angle.to(this.extension));
	}

	/** Retrieve trail as vector of trail points */
	public List<Vector2D> getTrail(final Vector2D start, final Vector2D end)
	{
		Pose2D poseStart = getClosestPose(start);
		Pose2D poseEnd = getClosestPose(end);
		double distance = poseStart.getDistanceTo(poseEnd);
		double angle = poseStart.getAngle().radians();
		double length = 0.1 * distance;
		Vector2D offset = new Vector2D(Math.cos(angle) * length, Math.sin(angle) * length);
		Vector3D point3D = poseStart.getPosition();
		Vector2D point = new Vector2D(point3D.getX(), point3D.getY());
		ArrayList<Vector2D> points = new ArrayList<>();
		int i = 0;
		do {
			point = point.add(offset);
			points.add(point);
		} while (++i < 10);
		return points;
	}

	@Override
	public boolean equals(Object other)
	{
		Line2D otherLine = (Line2D) other;
		return this.start == otherLine.start && this.extension == otherLine.extension;
	}

	/** Validate that this instance represents a proper line. If not, make this line the x-axis. */
	private void validate()
	{
		if (this.extension.getX() == 0 && this.extension.getY() == 0) {
			this.extension = new Vector2D(0, 1);
		}
	}
}
